specreduce/extract.py

```diff
@@ -143,3 +145,108 @@ def _ap_weight_image(trace, width, disp_axis, crossdisp_axis, image_shape):
+
+        crossdisp_axis : int, optional
+            The index of the image's cross-dispersion axis. [default: 0]
+
+        bkgrd_prof : `~astropy.modeling.Model`
+            A model for the image's background flux.
+            [default: models.Polynomial1D(2)]
+
+        Returns
+        -------
+        spec_1d : `~specutils.Spectrum1D`
+            The final, Horne extracted 1D spectrum.
+        """
+        # co-add signal in each column
+        ncols = image.shape[crossdisp_axis]
+        xd_pixels = np.arange(ncols) # y plot dir / x spec dir
+        coadd = image.sum(axis=disp_axis) / ncols
+
+        # fit source profile, using Gaussian model as a template
+        # NOTE: could add argument for users to provide their own model
+        gauss_prof = models.Gaussian1D(amplitude=coadd.max(),
+                                       mean=coadd.argmax(), stddev=2)
+
+        # Fit extraction kernel to column with combined gaussian/bkgrd model
+        ext_prof = gauss_prof + bkgrd_prof
+        fitter = fitting.LevMarLSQFitter()
+        fit_ext_kernel = fitter(ext_prof, xd_pixels, coadd)
+
+        # generate 1D spectrum, column by column
+        extracted = []
+        for col_pix in range(image.shape[disp_axis]):
+            # set gaussian model's mean as column's corresponding trace value
+            fit_ext_kernel.mean_0 = trace_object.trace[col_pix]
+            # kernel_col.stddev_0 = self.fwhm_fit(x)
+            # NOTE: support for variable FWHMs forthcoming
+
+            # fit compound model to column
+            kernel_vals = fit_ext_kernel(xd_pixels)
+            # NOTE: support user-provided kernels? would require different handling
+            # NOTE: allow smoothing in x direction via boxcar/x axis collapse/etc?
+
+            # fetch matching columns from original and variance images
+            image_col = image[:, col_pix]
+            variance_col = variance[:, col_pix]
+
+            # calculate kernel normalization
+            g_x = np.ma.sum(kernel_vals**2 / variance_col)
+            if np.ma.is_masked(g_x):
+                continue
+
+            # sum by column weights
+            weighted_col = np.ma.divide(image_col * kernel_vals, variance_col)
+            result = np.ma.sum(weighted_col) / g_x
+
+            # multiply kernel normalization into the extracted signal
+            result *= (fit_ext_kernel.amplitude_0
+                       * fit_ext_kernel.stddev_0 * np.sqrt(2*np.pi))
+            extracted.append(result)
+
+        # convert the extraction to a Spectrum1D object
+        pixels = np.arange(image.shape[disp_axis]) * u.pix
+        spec_1d = Spectrum1D(spectral_axis=pixels,
+                             flux=extracted * getattr(image, 'unit', u.DN))
+
+        return spec_1d
+
+
+@dataclass
+class OptimalExtract(HorneExtract):
+    """
+    Perform a Horne extraction (a.k.a. optimal) on a region in a
+    two-dimensional spectrum.
+    """
```

Comment:
if this isn't included, will the docstring still inherit from `HorneExtract`?  If so, might be nice to not duplicate just so they don't accidentally get out of sync.  Or maybe should just link to `HorneExtract` and say its an alias?

Reply:
i would mention in this docstring that this is just an alias for its parent class. sphinx `automodapi` will include that and create links to parent classes. not exactly sure how ipython docs and autocomplete do it, but i think they will follow up the inheritance tree.

Reply:
I don't see `HorneExtract`'s docstring when I remove `OptimalExtract`'s and import it in JupyterLab.

A quick and dirty solution is to include a note that it's an alias as Tim suggests and then say `__doc__ += HorneExtract.__doc__`, though I'm not sure how that would affect Sphinx.

Reply:
I got `automodapi` and `numpydoc` working well enough to find that the strategy mentioned in my previous comment works for the class-level docstring.

For `OptimalExtract.__call__()`, I see a brief method summary but not the full docstring. I'm not sure if this is user error (I'm seeing a lot of other warnings and am not that familiar with sphinx) or if there's more to do to make sure inherited class methods are properly documented.

Reply:
i am not a sphinx expert, either, but i will ping some people who are to get some guidance. it's hard enough to keep docs up to date with code, let alone multiple copies of basically the same docs...